Various policy improvements

tgraf opened the pull request:
<a href='#crh-start'></a><a href='#crh-data-%7B%22processed%22%3A%20%5B%22https%3A//github.com/cilium/cilium/pull/306%23discussion_r105389426%22%2C%20%22https%3A//github.com/cilium/cilium/pull/306%23discussion_r105389676%22%2C%20%22https%3A//github.com/cilium/cilium/pull/306%23discussion_r105390049%22%2C%20%22https%3A//github.com/cilium/cilium/pull/306%23discussion_r105390326%22%2C%20%22https%3A//github.com/cilium/cilium/pull/306%23discussion_r105390588%22%2C%20%22https%3A//github.com/cilium/cilium/pull/306%23discussion_r105397531%22%2C%20%22https%3A//github.com/cilium/cilium/pull/306%23discussion_r105397786%22%2C%20%22https%3A//github.com/cilium/cilium/pull/306%23discussion_r105397890%22%2C%20%22https%3A//github.com/cilium/cilium/pull/306%23discussion_r105398304%22%2C%20%22https%3A//github.com/cilium/cilium/pull/306%23discussion_r105398613%22%2C%20%22https%3A//github.com/cilium/cilium/pull/306%23discussion_r105398919%22%2C%20%22https%3A//github.com/cilium/cilium/pull/306%23discussion_r105401578%22%2C%20%22https%3A//github.com/cilium/cilium/pull/306%23discussion_r105599364%22%2C%20%22https%3A//github.com/cilium/cilium/pull/306%23discussion_r105660321%22%5D%2C%20%22comments%22%3A%20%7B%22Pull%206927602d2b73c89de878c5a23486231c8ac2ead2%20pkg/labels/labels.go%2099%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/cilium/cilium/pull/306%23discussion_r105390588%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Maybe%20we%20should%20pass%20an%20argument%20to%20%60AbsoluteKey%60%20with%20the%20list%20of%20valid%20prefixes%3F%20Depends%20if%20the%20caller%20is%20always%20the%20daemon.%22%2C%20%22created_at%22%3A%20%222017-03-10T12%3A53%3A31Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars3.githubusercontent.com/u/5714066%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/aanm%22%7D%7D%2C%20%7B%22body%22%3A%20%22With%20this%20change.%20Only%20the%20%60root.%60%20prefix%20is%20a%20valid%20prefix.%20I%20will%20remove%20this%20FIXME.%5Cr%5Cn%5Cr%5CnWhat%20we%20can%20talk%20about%20though%20is%20validating%20absolute%20label%20names%20as%20they%20are%20loaded%20as%20policy%20and%20then%20reject%20them.%20This%20way%20it%20will%20become%20trivial%20to%20notice%20if%20a%20policy%20contains%20elements%20which%20are%20not%20covered%20by%20the%20configured%20prefix.%22%2C%20%22created_at%22%3A%20%222017-03-10T13%3A44%3A22Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars2.githubusercontent.com/u/1417913%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/tgraf%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20pkg/labels/labels.go%3AL203-237%22%7D%2C%20%22Pull%206927602d2b73c89de878c5a23486231c8ac2ead2%20daemon/policy_test.go%2066%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/cilium/cilium/pull/306%23discussion_r105390326%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Shouldn%27t%20we%20add%20%60root%60%20has%20valid%20label%20prefixes%3F%22%2C%20%22created_at%22%3A%20%222017-03-10T12%3A51%3A27Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars3.githubusercontent.com/u/5714066%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/aanm%22%7D%7D%2C%20%7B%22body%22%3A%20%22Not%20sure%20I%20understand.%20These%20tests%20validate%20proper%20label%20resolving.%20%60root.%60%20is%20the%20absolute%20name%20of%20these%20labels.%20The%20user%20will%20never%20see%20it%20this%20way.%22%2C%20%22created_at%22%3A%20%222017-03-10T13%3A42%3A15Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars2.githubusercontent.com/u/1417913%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/tgraf%22%7D%7D%2C%20%7B%22body%22%3A%20%22Ah%2C%20so%20the%20user%20type%20%60docker%20--label%20%5C%22Prod%5C%22%60%20and%20the%20label%20will%20actually%20be%20%60root.Prod%60%3F%22%2C%20%22created_at%22%3A%20%222017-03-10T13%3A47%3A50Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars3.githubusercontent.com/u/5714066%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/aanm%22%7D%7D%2C%20%7B%22body%22%3A%20%22Yes%22%2C%20%22created_at%22%3A%20%222017-03-10T14%3A02%3A32Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars2.githubusercontent.com/u/1417913%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/tgraf%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20daemon/policy_test.go%3AL36-52%22%7D%2C%20%22Pull%2068f0f99617222fd0d6b9b77a874231ca45ed51b1%20daemon/main.go%2012%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/cilium/cilium/pull/306%23discussion_r105389676%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Should%20we%20also%20add%20this%20options%20on%20%60cilium%20client%60%3F%22%2C%20%22created_at%22%3A%20%222017-03-10T12%3A46%3A25Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars3.githubusercontent.com/u/5714066%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/aanm%22%7D%7D%2C%20%7B%22body%22%3A%20%22Why%20do%20we%20need%20it%20for%20the%20client%3F%20I%20would%20rather%20not%20complicate%20this%20unless%20required.%20Restarting%20the%20daemon%20is%20possible%20without%20losing%20state%20so%20unless%20we%20get%20a%20direct%20request%20for%20making%20this%20changeable%20at%20runtime%20I%20don%27t%20see%20a%20need%20for%20it.%22%2C%20%22created_at%22%3A%20%222017-03-10T13%3A40%3A15Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars2.githubusercontent.com/u/1417913%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/tgraf%22%7D%7D%2C%20%7B%22body%22%3A%20%22%40aanm%20Ping%20on%20this%20comment.%20Please%20clarify%20whether%20you%20still.%22%2C%20%22created_at%22%3A%20%222017-03-13T07%3A06%3A32Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars2.githubusercontent.com/u/1417913%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/tgraf%22%7D%7D%2C%20%7B%22body%22%3A%20%22It%20was%20just%20a%20thought%2C%20let%27s%20keep%20as%20is%20and%20change%20it%20in%20the%20future%20if%20needed.%22%2C%20%22created_at%22%3A%20%222017-03-13T13%3A35%3A48Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars3.githubusercontent.com/u/5714066%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/aanm%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20daemon/main.go%3AL107-115%22%7D%2C%20%22Pull%2068f0f99617222fd0d6b9b77a874231ca45ed51b1%20pkg/labels/filter.go%208%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/cilium/cilium/pull/306%23discussion_r105389426%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Should%20we%20validate%20the%20source%3F%20What%20happens%20if%20user%20misspells%20%60cilium%60%20for%20example%3F%22%2C%20%22created_at%22%3A%20%222017-03-10T12%3A44%3A51Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars3.githubusercontent.com/u/5714066%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/aanm%22%7D%7D%2C%20%7B%22body%22%3A%20%22Then%20the%20label%20will%20not%20match.%20How%20do%20you%20know%20the%20valid%20label%20sources%20at%20this%20point.%20It%20could%20be%20anything.%20As%20we%20make%20the%20label%20retrieving%20more%20modular%2C%20we%20will%20have%20no%20idea%20about%20all%20possible%20sources%20of%20labels.%22%2C%20%22created_at%22%3A%20%222017-03-10T13%3A46%3A10Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars2.githubusercontent.com/u/1417913%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/tgraf%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20pkg/labels/filter.go%3AL34-65%22%7D%2C%20%22Pull%2057ef910e2cb2dd9e8564b4b771cde664884c99d5%20pkg/labels/filter.go%206%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/cilium/cilium/pull/306%23discussion_r105390049%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Why%20ID%3F%22%2C%20%22created_at%22%3A%20%222017-03-10T12%3A49%3A14Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars3.githubusercontent.com/u/5714066%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/aanm%22%7D%7D%2C%20%7B%22body%22%3A%20%22%60id.%60%20is%20a%20short%20prefix%20that%20we%20can%20use%20in%20examples.%20%60id.frontend%60%20talks%20to%20%60id.backend%60%22%2C%20%22created_at%22%3A%20%222017-03-10T13%3A41%3A44Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars2.githubusercontent.com/u/1417913%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/tgraf%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20pkg/labels/filter.go%3AL69-79%22%7D%7D%7D'></a>
<a href='https://www.codereviewhub.com/'><img src='http://www.codereviewhub.com/site/github-bar.png' height=40></a>
- [x] <a href='#crh-comment-Pull 68f0f99617222fd0d6b9b77a874231ca45ed51b1 pkg/labels/filter.go 8'></a> <img src='http://www.codereviewhub.com/site/github-remaining.png' height=16 width=60>&nbsp;<b><a href='https://github.com/cilium/cilium/pull/306#discussion_r105389426'>File: pkg/labels/filter.go:L34-65</a></b>
- <a href='https://github.com/aanm'><img border=0 src='https://avatars3.githubusercontent.com/u/5714066?v=3' height=16 width=16></a> Should we validate the source? What happens if user misspells `cilium` for example?
- <a href='https://github.com/tgraf'><img border=0 src='https://avatars2.githubusercontent.com/u/1417913?v=3' height=16 width=16></a> Then the label will not match. How do you know the valid label sources at this point. It could be anything. As we make the label retrieving more modular, we will have no idea about all possible sources of labels.
- [x] <a href='#crh-comment-Pull 68f0f99617222fd0d6b9b77a874231ca45ed51b1 daemon/main.go 12'></a> <img src='http://www.codereviewhub.com/site/github-remaining.png' height=16 width=60>&nbsp;<b><a href='https://github.com/cilium/cilium/pull/306#discussion_r105389676'>File: daemon/main.go:L107-115</a></b>
- <a href='https://github.com/aanm'><img border=0 src='https://avatars3.githubusercontent.com/u/5714066?v=3' height=16 width=16></a> Should we also add this options on `cilium client`?
- <a href='https://github.com/tgraf'><img border=0 src='https://avatars2.githubusercontent.com/u/1417913?v=3' height=16 width=16></a> Why do we need it for the client? I would rather not complicate this unless required. Restarting the daemon is possible without losing state so unless we get a direct request for making this changeable at runtime I don't see a need for it.
- <a href='https://github.com/tgraf'><img border=0 src='https://avatars2.githubusercontent.com/u/1417913?v=3' height=16 width=16></a> @aanm Ping on this comment. Please clarify whether you still.
- <a href='https://github.com/aanm'><img border=0 src='https://avatars3.githubusercontent.com/u/5714066?v=3' height=16 width=16></a> It was just a thought, let's keep as is and change it in the future if needed.
- [x] <a href='#crh-comment-Pull 57ef910e2cb2dd9e8564b4b771cde664884c99d5 pkg/labels/filter.go 6'></a> <img src='http://www.codereviewhub.com/site/github-remaining.png' height=16 width=60>&nbsp;<b><a href='https://github.com/cilium/cilium/pull/306#discussion_r105390049'>File: pkg/labels/filter.go:L69-79</a></b>
- <a href='https://github.com/aanm'><img border=0 src='https://avatars3.githubusercontent.com/u/5714066?v=3' height=16 width=16></a> Why ID?
- <a href='https://github.com/tgraf'><img border=0 src='https://avatars2.githubusercontent.com/u/1417913?v=3' height=16 width=16></a> `id.` is a short prefix that we can use in examples. `id.frontend` talks to `id.backend`
- [x] <a href='#crh-comment-Pull 6927602d2b73c89de878c5a23486231c8ac2ead2 daemon/policy_test.go 66'></a> <img src='http://www.codereviewhub.com/site/github-remaining.png' height=16 width=60>&nbsp;<b><a href='https://github.com/cilium/cilium/pull/306#discussion_r105390326'>File: daemon/policy_test.go:L36-52</a></b>
- <a href='https://github.com/aanm'><img border=0 src='https://avatars3.githubusercontent.com/u/5714066?v=3' height=16 width=16></a> Shouldn't we add `root` has valid label prefixes?
- <a href='https://github.com/tgraf'><img border=0 src='https://avatars2.githubusercontent.com/u/1417913?v=3' height=16 width=16></a> Not sure I understand. These tests validate proper label resolving. `root.` is the absolute name of these labels. The user will never see it this way.
- <a href='https://github.com/aanm'><img border=0 src='https://avatars3.githubusercontent.com/u/5714066?v=3' height=16 width=16></a> Ah, so the user type `docker --label "Prod"` and the label will actually be `root.Prod`?
- <a href='https://github.com/tgraf'><img border=0 src='https://avatars2.githubusercontent.com/u/1417913?v=3' height=16 width=16></a> Yes
- [x] <a href='#crh-comment-Pull 6927602d2b73c89de878c5a23486231c8ac2ead2 pkg/labels/labels.go 99'></a> <img src='http://www.codereviewhub.com/site/github-remaining.png' height=16 width=60>&nbsp;<b><a href='https://github.com/cilium/cilium/pull/306#discussion_r105390588'>File: pkg/labels/labels.go:L203-237</a></b>
- <a href='https://github.com/aanm'><img border=0 src='https://avatars3.githubusercontent.com/u/5714066?v=3' height=16 width=16></a> Maybe we should pass an argument to `AbsoluteKey` with the list of valid prefixes? Depends if the caller is always the daemon.
- <a href='https://github.com/tgraf'><img border=0 src='https://avatars2.githubusercontent.com/u/1417913?v=3' height=16 width=16></a> With this change. Only the `root.` prefix is a valid prefix. I will remove this FIXME.
What we can talk about though is validating absolute label names as they are loaded as policy and then reject them. This way it will become trivial to notice if a policy contains elements which are not covered by the configured prefix.


<a href='https://www.codereviewhub.com/cilium/cilium/pull/306?mark_as_completed=1'><img src='http://www.codereviewhub.com/site/github-mark-as-completed.png' height=26></a>&nbsp;<a href='https://www.codereviewhub.com/cilium/cilium/pull/306?approve=1'><img src='http://www.codereviewhub.com/site/github-approve.png' height=26></a>&nbsp;<a href='https://github.com/cilium/cilium/pull/306'><img src='http://www.codereviewhub.com/site/github-refresh.png' height=26></a>
<a href='#crh-end'></a>

aanm commented on pkg/labels/filter.go:
```diff
@@ -34,13 +34,31 @@ type LabelPrefix struct {
+	labelPrefix := LabelPrefix{}
+	t := strings.SplitN(label, ":", 2)
+	if len(t) > 1 {
+		labelPrefix.Source = t[0]
```
Should we validate the source? What happens if user misspells `cilium` for example?

tgraf replied:
Then the label will not match. How do you know the valid label sources at this point. It could be anything. As we make the label retrieving more modular, we will have no idea about all possible sources of labels.

aanm commented on daemon/main.go:
```diff
@@ -106,6 +107,8 @@ func init() {
 	flags.BoolVar(&config.KeepConfig, "keep-config", false,
 		"When restoring state, keeps containers' configuration in place")
 	flags.StringVar(&labelPrefixFile, "label-prefix-file", "", "File with valid label prefixes")
+	flags.StringSliceVar(&validLabels, "labels", []string{},
```
Should we also add this options on `cilium client`? Plus, add and remove valid labels without restarting cilium.

tgraf replied:
Why do we need it for the client? I would rather not complicate this unless required. Restarting the daemon is possible without losing state so unless we get a direct request for making this changeable at runtime I don't see a need for it.

tgraf replied:
@aanm Ping on this comment. Please clarify whether you still.

aanm replied:
It was just a thought, let's keep as is and change it in the future if needed.

aanm commented on pkg/labels/filter.go:
```diff
@@ -69,16 +69,10 @@ func DefaultLabelPrefixCfg() *LabelPrefixCfg {
 			{
-				Prefix: common.GlobalLabelPrefix,
-				Source: common.ReservedLabelSource,
+				Prefix: "id.",
```
Why ID?

tgraf replied:
`id.` is a short prefix that we can use in examples. `id.frontend` talks to `id.backend`

aanm commented on daemon/policy_test.go:
```diff
@@ -36,71 +36,16 @@ var (
-	lblBar := labels.NewLabel("io.cilium.bar", "", common.CiliumLabelSource)
-	lblJoe := labels.NewLabel("io.cilium.user", "joe", common.CiliumLabelSource)
-	lblPete := labels.NewLabel("io.cilium.user", "pete", common.CiliumLabelSource)
+	lblProd := labels.NewLabel("root.Prod", "", common.CiliumLabelSource)
```
Shouldn't we add `root` has valid label prefixes?

tgraf replied:
Not sure I understand. These tests validate proper label resolving. `root.` is the absolute name of these labels. The user will never see it this way.

aanm replied:
Ah, so the user type `docker --label "Prod"` and the label will actually be `root.Prod`?

tgraf replied:
Yes

aanm commented on pkg/labels/labels.go:
```diff
@@ -186,59 +203,34 @@ func (l *Label) Matches(target *Label) bool {
-		return l.absKey
+	if l.absKey == "" {
+		// Never translate using an owner if a reserved label
+		// FIXME: the HasPrefix should be using daemon.config.ValidLabelPrefixes
```
Maybe we should pass an argument to `AbsoluteKey` with the list of valid prefixes? Depends if the caller is always the daemon.

tgraf replied:
With this change. Only the `root.` prefix is a valid prefix. I will remove this FIXME.

What we can talk about though is validating absolute label names as they are loaded as policy and then reject them. This way it will become trivial to notice if a policy contains elements which are not covered by the configured prefix.

aanm reviewed (changes requested):
We also need a policy doc with all the details from all the commits.

tgraf committed:
policy: Rename PolicyRuleRequires and PolicyRuleConsumers

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
policy: Continue evaluating policy on single consumer eval failure

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
daemon: Add `--labels` option to add additional allowed label prefixes

In addition to the file based label prefix configuration, allow
specifying additional label prefixes to be used when determing
the identity of an endpoint.

Example:
  cilium-agent --labels org.corp.id.

Will taken into account all labels which start with the prefix
"org.corp.id." for identification purposes, i.e.:

  docker run -l org.corp.id.team=foo [...]
  docker run -l org.corp.id.team=bar [...]

Will result in both containers to receive distinct identities and
allow specyfing policy based on these labels.

This commit also extends the filter logic to make the source portion
of the label prefix optional. If ommited, the label prefix will match
all alls, e.g.:

  cilium-agent --labels docker:org.corp.id.

will only match docker runtime labels with the prefix "org.corp.id."

wheras:

  cilium-gent --labels org.corp.id.

will match labels from all orchestration systems (k8s, docker, ...)

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
labels: Whitelist "id." prefix by default

So far, the label prefix "io.cilium" has been whitelisted by default.
Additionally whitelist "id.". This can be overwritten using the
--label-prefix-file option.

This allows the following construct to work out of the box:
  docker run -l id.frontend [....]
  docker run -l id.backend [...]

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
policy: Restrict node merging

Merging of nodes has been allowed unconditionally with the expectations
that users know what they are doing. Merging of nodes is perfectly fine
up to the point where rules contain deny and always-accept actions in
which case the order of the rules start to matter.

Prevent accidental mistakes by refusing to merge nodes which are not
safely mergegable.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
policy: Remove requirement of single shared label root prefix

Up to now, we required all policy node names to share a single
prefix which defaulted to "io.cilium". This was configurable
but required a specific label addressing schema.

This commits changes this and puts all labels under a single root
label called "root". Rules attached to the root label applied to
all endpoints. The root node can have children, e.g.

          root
       io       org
  k8s     foo       corp
                 foo    bar

Signed-off-by: Thomas Graf <thomas@cilium.io>
